Stateful app backup policies

sahare opened the pull request:
The Policies available here provide backup and restore support for stateful applications running on managed clusters or hub. Velero is used to backup and restore applications data.
Policies:

oadp-hdr-app-install installs oadp and configure connection to the storage on managed clusters
oadp-hdr-app-backup backups applications on managed clusters
oadp-hdr-app-restore applications on managed clusters
The configmap hdr-app-configmap in the PR sets configuration options for the backup storage location, for the backup schedule backing up applications, and for the restore resource used to restore applications backups.

https://issues.redhat.com/browse/ACM-4708

birsanv commented:
this is a continuation of https://github.com/open-cluster-management-io/policy-collection/pull/409

dhaiducek reviewed (commented):
I'm discussing some things about this PR with our team before approving it. In particular, PlacementRule is deprecated, so while existing policies in the collection use PlacementRule, it might be good to go ahead and update this policy to use Placement now rather than having to revisit in the future. I'll confirm this, though.

dhaiducek commented on community/CM-Configuration-Management/acm-app-pv-backup/resources/policy-sets/acm-app-backup-policy-set.yaml:
```diff
@@ -0,0 +1,45 @@
+    - oadp-hdr-app-install
+    - oadp-hdr-app-backup
+---
+apiVersion: apps.open-cluster-management.io/v1
+kind: PlacementRule
+metadata:
+  name: acm-app-backup-placement
+spec:
+  clusterConditions: []
+  clusterSelector:
+    matchExpressions:
+      - key: acm-pv-dr
+        operator: In
+        values:
+          - backup
```
Please update this to use Placement since PlacementRule is deprecated.

dhaiducek reviewed (changes requested):
Please update all PlacementRules (and PlacementBinding references in conjunction) to use Placement now that PlacementRule is deprecated.

mprahl commented on community/CM-Configuration-Management/acm-app-pv-backup/resources/policies/oadp-hdr-app-backup.yaml:
```diff
@@ -0,0 +1,186 @@
+    name: oadp-hdr-app-install
+  disabled: false
+  policy-templates:
+    - objectDefinition:
```
I just came in to say that there are a lot of `fromConfigMap` calls to the same `ConfigMap` and those API calls are not cached unfortunately. One alternative is to use `object-templates-raw` in each `ConfigurationPolicy` which is just a field that allows templating as long as the final output is valid YAML following the syntax of `object-templates`.

So you could use `with` from go text/template to store the value of the `ConfigMap` using the `lookup` template function. Then at least it'd only be one API call for the `ConfigMap` per `ConfigurationPolicy`.

mprahl replied:
Note that object-templates-raw is only in ACM 2.7+.

birsanv replied:
@mprahl can you add a sample please for object-templates-raw usage

mprahl replied:
@birsanv not exactly your situation but here is an example:
```yaml
apiVersion: policy.open-cluster-management.io/v1
kind: Policy
metadata:
  name: label-configmaps
  namespace: policies
  annotations:
    policy.open-cluster-management.io/standards: NIST SP 800-53
    policy.open-cluster-management.io/categories: CM Configuration Management
    policy.open-cluster-management.io/controls: CM-2 Baseline Configuration
spec:
  disabled: false
  policy-templates:
    - objectDefinition:
        apiVersion: policy.open-cluster-management.io/v1
        kind: ConfigurationPolicy
        metadata:
          name: label-configmaps
        spec:
          remediationAction: enforce
          severity: low
          object-templates-raw: |
            {{- range (lookup "v1" "ConfigMap" "my-configmaps" "").items }}
            {{- if and .data.model (contains "Mustang" .data.model) }}
            - complianceType: musthave
              objectDefinition:
                kind: ConfigMap
                apiVersion: v1
                metadata:
                  name: {{ .metadata.name }}
                  namespace: {{ .metadata.namespace }}
                  labels:
                    ford.com/model: Mustang
            {{- end }}
            {{- end }}
```

birsanv replied:
@mprahl this is quite a big change, is this needed to approve the PR ?

I could see a basic sample of the object-templates-raw with the documentation but I am not clear what changes need to be done to replace object-templates with object-templates-raw. 

>>So you could use with from go text/template to store the value of the ConfigMap using the lookup template function. Then at least it'd only be one API call for the ConfigMap per ConfigurationPolicy

birsanv replied:
@mprahl I understand what the ask is, no need for a sample
Looking into making the requested changes, if they are needed for the PR to be approved..

birsanv replied:
There are a few issues @mprahl when trying to use object-templates-raw + lookup instead of   object-templates + fromConfigMap

1. Since the user can create the policy on the hub under any namespace, the configmap used by the policy doesn't have a fixed namespace ( the configmap ns would be policy's ns ). So in the object-templates, the  fromConfigMap uses "" for the ns , for example '{{hub fromConfigMap "" "hdr-app-configmap" "backupNS" hub}}'.  This defaults to the policy namespace

If I change the  object-templates with  object-templates-raw, this is no longer working, I have to specify the namespace for the configmap

I have to change 
          ```
object-templates:
            - complianceType: mustnothave
              objectDefinition:
                apiVersion: velero.io/v1
                kind: Schedule
                metadata:
                  namespace: '{{hub fromConfigMap "" "hdr-app-configmap" "backupNS" hub}}'
                  labels:
                    cluster-id: '{{ fromClusterClaim "id.openshift.io" }}'
                    cluster-name: '{{ fromClusterClaim "name" }}'              
                status:
                  phase: FailedValidation                
          remediationAction: inform
          severity: high
```


with 

```
          object-templates-raw: |
            - complianceType: musthave
              objectDefinition:
                apiVersion: velero.io/v1
                kind: Schedule
                metadata:
                  namespace: '{{ (lookup "v1" "ConfigMap" "**app-policies**" "hdr-app-configmap").data.backupNS }}'
                  labels:
                    cluster-id: '{{ fromClusterClaim "id.openshift.io" }}'
                    cluster-name: '{{ fromClusterClaim "name" }}'
                status:
                  phase: FailedValidation                
          remediationAction: inform
          severity: high
```

2.  The configmap has keys in this format a.b ( for example backup.volumeSnapshotLocation ) which are supported by the fromConfigMap but not by the lookup since I assumes it takes this as a  path

Item 2 can be fixed, rename all the properties and take out the .
How do I implement point 1. though with the lookup function ?

mprahl replied:
@birsanv if this is not an out of the box policy or commonly used policy, then the change is not required. If it will be commonly deployed, I want to minimize the impacts on the Kubernetes API.

For point 1, you can still use a mix of Hub templates and managed cluster side templates. That hasn't changed.

This untested, but this is along the lines of what I was thinking of:
```yaml
      object-templates-raw: |
        {{hub with $configMap := (lookup "v1" "ConfigMap" "" "hdr-app-configmap") hub}}
        - complianceType: musthave
          objectDefinition:
            apiVersion: velero.io/v1
            kind: Schedule
            metadata:
              namespace: '{{hub $configMapdata.data.backupNS hub}}'
              labels:
                cluster-id: '{{ fromClusterClaim "id.openshift.io" }}'
                cluster-name: '{{ fromClusterClaim "name" }}'
            status:
              phase: FailedValidation     
          {{hub end hub}}           
      remediationAction: inform
      severity: high
```

Also, in ACM 2.8+, you can now use the `.ManagedClusterLabels` variable in Hub templates to potentially avoid `fromClusterClaim` calls on the managed cluster side.

For point 2, you can use the `index` template function to access properties/keys that have characters such as `.` and `-`.:
https://pkg.go.dev/text/template#hdr-Functions

birsanv replied:
@mprahl thanks
I tried this template

```
    - objectDefinition:
        apiVersion: policy.open-cluster-management.io/v1
        kind: ConfigurationPolicy
        metadata:
          name: check-schedule-failed-validation-2
        spec:
          object-templates-raw: |
            {{hub with $configMap := (lookup "v1" "ConfigMap" "" "hdr-app-configmap") hub}}
            - complianceType: musthave
              objectDefinition:
                apiVersion: velero.io/v1
                kind: Schedule
                metadata:
                  namespace: '{{hub $configMapdata.data.backupNS hub}}'
                  labels:
                    cluster-id: '{{ fromClusterClaim "id.openshift.io" }}'
                    cluster-name: '{{ fromClusterClaim "name" }}'
                status:
                  phase: FailedValidation     
              {{hub end hub}}                
          remediationAction: inform
          severity: high
```

and get the following error

    policy.open-cluster-management.io/hub-templates-error: >-
      failed to parse the template JSON string
      {"apiVersion":"policy.open-cluster-management.io/v1","kind":"ConfigurationPolicy","metadata":{"name":"check-schedule-failed-validation-2"},"spec":{"object-templates-raw":"{{hub
      with $configMap := (lookup \"v1\" \"ConfigMap\" \"\"
      \"hdr-app-configmap\") hub}}\n- complianceType: musthave\n 
      objectDefinition:\n    apiVersion: velero.io/v1\n    kind: Schedule\n   
      metadata:\n      namespace: '{{hub $configMapdata.data.backupNS
      hub}}'\n      labels:\n        cluster-id: '{{ fromClusterClaim
      \"id.openshift.io\" }}'\n        cluster-name: '{{ fromClusterClaim
      \"name\" }}'\n    status:\n      phase: FailedValidation     \n  {{hub end
      hub}}                \n","remediationAction":"inform","severity":"high"}}:
      template: tmpl:6: unexpected "\\" in operand
  creationTimestamp: '2023-05-24T14:26:51Z'

I don't understand the error message.. I tried to look for similar doc or samples using {{hub with .. }}  but can't find something similar to debug the issue

I am actually wondering if this will solve my issue at point 1 after all ( the lookup asks for the configMap namespace to be specified , a "" value is not looking at the policy namespace on the hub ; whereas the fromConfigMap with an "" ns gets the configmap from the policy namespace )
Wouldn't 
`{{hub with $configMap := (lookup "v1" "ConfigMap" "" "hdr-app-configmap") hub}} ` 

produce the same result as

`namespace: '{{ (lookup "v1" "ConfigMap" "" "hdr-app-configmap").data.backupNS }}'`

? 

The lookup above works only if I enter the ns name

`namespace: '{{ (lookup "v1" "ConfigMap" "app-policies" "hdr-app-configmap").data.backupNS }}'`

mprahl replied:
To answer your question, `{{` is on the managed cluster side. So `app-policies` would only be there if the policy applied to a self-managed Hub. Hub templates get executed before the policy is copied to the managed cluster.

sahare commented:
@mprahl Please review, pushed some changes trying to address your request as follows:

- renamed properties that have .
- replaced fromConfigMap with lookup
- used the object-templates-raw in places where it didn't throw an exception; basically simple templates, mostly the install policy

mprahl reviewed (dismissed):
Looks great! Thanks for making the changes.

/hold

sahare commented:
/retest

sahare commented:
@mprahl We are done with testing, could you please re-approve and merge. Thanks!

sahare commented:
/remove-hold

dhaiducek reviewed (approved):
/hold in case you want to address the comment

Looks good! Thanks for this contribution, and thanks for your patience and iterations through the PlacementRule deprecation!

birsanv commented:
@dhaiducek the {{hub (printf "%s" .ManagedClusterName) hub}}") is outside the end block onpurpose, because it is not processed properly if it's inside .. :) you can verify, I get a nil value

sahare commented:
/remove-hold
